macro/src/lib.rs

```diff
@@ -96,6 +96,115 @@ fn param_conversion(
+            Some(ret_ty) => {
+                let conv_ret_ty = ret_ty.to_syn();
+                quote!( -> #conv_ret_ty)
+            }
+            None => {
+                quote! {}
+            }
+        };
+        param_types.insert(0, syn::parse_quote!(*const c_void));
+        method_sigs.push(quote!(
+            pub #method_name: unsafe extern "C" fn (#(#param_types),*) #return_tokens,
+
+        ));
+    }
+    syn::parse_quote!(
+        #[repr(C)]
+        pub struct #custom_trait_vtable_type {
+            #(#method_sigs)*
+        }
+    )
+}
+
+fn gen_custom_trait_impl(custom_trait: &ast::Trait, custom_trait_struct_name: &Ident) -> Item {
+    let mut methods: Vec<Item> = vec![];
+    for m in &custom_trait.fcts {
+        let param_names: Vec<proc_macro2::TokenStream> = m
+            .params
+            .iter()
+            .map(|p| {
+                let p_name = &p.name;
+                quote! {, #p_name}
+            })
+            .collect();
+        let mut param_names_and_types: Vec<proc_macro2::TokenStream> = m
+            .params
+            .iter()
+            .map(|p| {
+                let p_ty = param_ty(&p.ty);
+                let p_name = &p.name;
+                quote!(#p_name : #p_ty)
+            })
+            .collect();
+        let method_name = &m.name;
+        let (return_tokens, end_token) = match &m.output_type {
+            Some(ret_ty) => {
+                let conv_ret_ty = ret_ty.to_syn();
+                (quote!( -> #conv_ret_ty), quote! {})
+            }
+            None => (quote! {}, quote! {;}),
+        };
+        if let Some(self_param) = &m.self_param {
+            let mut self_modifier = quote! {};
+            if let Some((lifetime, mutability)) = &self_param.reference {
+                let lifetime_mod = if *lifetime == ast::Lifetime::Anonymous {
+                    quote! { & }
+                } else {
+                    let prime = "'".to_string();
+                    let lifetime = lifetime.to_syn();
+                    quote! { & #prime#lifetime }
+                };
+                let mutability_mod = if *mutability == ast::Mutability::Mutable {
+                    quote! {mut}
+                } else {
+                    quote! {}
+                };
+                self_modifier = quote! { #lifetime_mod #mutability_mod }
+            }
+            param_names_and_types.insert(0, quote!(#self_modifier self));
+        }
+
+        let lifetimes = {
+            let lifetime_env = &m.lifetimes;
+            if lifetime_env.is_empty() {
+                quote! {}
+            } else {
+                quote! { <#lifetime_env> }
+            }
+        };
+        let runner_method_name =
+            Ident::new(&format!("run_{}_callback", method_name), Span::call_site());
+        methods.push(syn::Item::Fn(syn::parse_quote!(
+            fn #method_name#lifetimes (#(#param_names_and_types),*) #return_tokens {
+                unsafe {
```

Comment:
param_conversions may need to be applied here as well, I think

Reply:
(see what happens if you have a trait method with `&[u8]`: it _should_ get converted to DiplomatSlice and generate an `.into()` somewhere)

Reply:
Ah you're right! This needs to be handled the same as callback params.